### PR TITLE
LPS-130706 Migrate asset categories selection modals

### PR DESCRIPTION
**Pull request by markocikos**
JIRA: https://issues.liferay.com/browse/LPS-130706

@jonmak08 @kresimir-coko 

In this PR, we are migrating asset categories selection modals, a part of asset-related modals. It contains all remaining code from https://github.com/liferay-frontend/liferay-portal/pull/951, as well as part of code from https://github.com/liferay-frontend/liferay-portal/pull/967.

After this PR, https://github.com/liferay-frontend/liferay-portal/pull/967 should contain only changes on content-dashboard module. That PR can be on hold until this one is merged.

I'm setting this as draft:
- it has a migration type that we've not done before
- it modifies risky `asset` module, we are pausing all PRs on it until the big release

The code for vocabulary deletion is the same as in https://github.com/liferay-frontend/liferay-portal/pull/951. See comments on it there. New code, related to category selection, is in a separate commit, https://github.com/liferay-frontend/liferay-portal/commit/f7b18df3d610959b2d2565f2ac99e1b2f5aca1f3 . Here are notes on that code:
- I renamed `ItemSelectorDetails` to `ItemSelectorAddCategory`, as it is very much specific to this case.
- I moved all code related to 'add category' feature from more general `SelectCategory` to `ItemSelectorAddCategory`. @jonmak08 When you modified content dashboard in https://github.com/liferay-frontend/liferay-portal/pull/967, there was no need for completely unrelated code to break. This way we will have better separation.
- We are adding a wrapper around buttons. This is to match Clay buttons markup.
- The existing solution, as well as the solution in this PR, contains a lot of hacky code, where we directly modify modal buttons and title. This may be a case for `useLiferayState`, where it would contain modal buttons and title. I am not sure if we can sync state in and out of `iframe` with `useLiferayState`. In any case, it would require a large restructure of both modal util and this solution, beyond the scope of this epic. I am not sure how often we directly modify modal frame from inside. If often, it may be something to have in mind @wincent  @jbalsas 

Test case 1:
1. Categorization > Categories > create a Vocabulary
2. Test deleting a vocabulary (see gif)
after2

Test case 2:
1. Have a Vocabulary
2. Web Content > new Basic Web Content > 'Select' next to Vocabulary
3. Test selecting categories (see gif)
4. Test creating a new category (see gif)
after1

### CODE REVIEW

**Comment by wincent** [comment]
> * The existing solution, as well as the solution in this PR, contains a lot of hacky code, where we directly modify modal buttons and title. This may be a case for `useLiferayState`, where it would contain modal buttons and title. I am not sure if we can sync state in and out of `iframe` with `useLiferayState`.

Not directly, but I would think it should be possible at least for the iframe to do stuff like `window.parent.Liferay.State.subscribe(someAtom, someCallback)`. But there are a couple of gotchas.

1. This assumes the content window of the iframe is same host, protocol, port etc (something we can expect to be true for the most part, if not always), and that `someCallback()` is going to work when called from the parent window (eg. without having to spell it out as `document.getElementById('theFrame').contentWindow.someReference.to.someCallback()`. Untested.
2. This assumes we have access to `someAtom` from the iframe, which we don't, unless we passed it down or assigned it some way previously (because if the iframe gets it via `Liferay.Loader.require` it's actually going to get its own copy of it). Again, untested; there might be some edge cases about object identity if you try to pass stuff down this way.

Of course, there is the `postMessage` API which is a bit less "sneaky" when it comes to communicating between frames, but it has obvious limitations. You can only pass "structured clonable" objects via `postMessage`, for example, and our global state can contain _anything_. So we might want to consider building something on top of the `Liferay.State.__unsafe__` layer, which would allow us to reference atoms and selectors by their string keys without actually having to load their concrete instances. You'd lose the type safety, of course, but it would at least cover a pretty common use case in Liferay DXP (communication between modal and parent).

**Comment by markocikos** [comment]
I pushed a [fix](https://github.com/liferay-frontend/liferay-portal/pull/976/commits/712fe1cb133d2e70c618d79e4c18fb0b6419a05e
) for `propTypes` warning that `addCategoryURL` is required. 

screen

It is not, some category selection modals do not have the 'add category' feature.

I was testing the CI failure in https://testray.liferay.com/reports/production/logs/2021-04/test-1-8/test-portal-acceptance-pullrequest(master)/7224/functional-tomcat90-mysql57-jdk8/0/4/LocalFile.AssetPublisherWithWebContent_ViewWebContentViaAPDeleteCategory/summary.html.gz. It looks to me like the CI test is outdated, looking for card classes instead of tree selector classes. We can revisit this after this PR graduates from being draft, after the release.

File: modules/apps/asset/asset-taglib/src/main/resources/META-INF/resources/asset_categories_selector/AssetVocabularyCategoriesSelector.es.js [hunk 1]
```diff
@@ -140,40 +140,35 @@ function AssetVocabulariesCategoriesSelector({
...
 		});
 
-		const itemSelectorDialog = new ItemSelectorDialog({
+		openSelectionModal({
 			buttonAddLabel: Liferay.Language.get('done'),
-			dialogClasses: 'modal-lg',
-			eventName,
+			multiple: true,
+			onSelect: (selectedItems) => {
+				if (selectedItems) {
```
Comment (kresimir-coko) [inline modules/apps/asset/asset-taglib/src/main/resources/META-INF/resources/asset_categories_selector/AssetVocabularyCategoriesSelector.es.js:147]: Hey @markocikos, smallest nitpick ever, but perhaps `selectedItems.length`?

**Review by kresimir-coko — APPROVED** [review]
Apart from the one inline comment looks not horrible

File: modules/apps/asset/asset-taglib/src/main/resources/META-INF/resources/asset_categories_selector/AssetVocabularyCategoriesSelector.es.js (hunk 1 above)
Comment (markocikos) [reply modules/apps/asset/asset-taglib/src/main/resources/META-INF/resources/asset_categories_selector/AssetVocabularyCategoriesSelector.es.js]: Updated to `if (selectedItems?.length) {`

**Comment by markocikos** [comment]
Forwarded to https://github.com/liferay-echo/liferay-portal/pull/4436

**Pull request closed without merging**